src/Swarm/Game/Step.hs

```diff
@@ -1412,6 +1383,68 @@ execConst c vs s k = do
     return . (if remTime <= 1 then id else Waiting (remTime + time)) $
       Out VUnit s (FImmediate wf rf : k)
 
+  lookInDirection ::
+    (Has (State GameState) sig m, Has (State Robot) sig m, Has (Error Exn) sig m) =>
+    Direction ->
+    m (V2 Int64, Maybe Entity)
+  lookInDirection d = do
+    loc <- use robotLocation
+    orient <- use robotOrientation
+    when (isCardinal d) $ hasCapabilityFor COrient (TDir d)
+    let nextLoc = loc ^+^ applyTurn d (orient ? zero)
+    (nextLoc,) <$> entityAt nextLoc
+
+  -- Find out the required devices for running the command on the
+  -- target robot - this is common for 'Build' and 'Reprogram'.
+  --
+  -- Note that _capCtx must be empty: at least at the
+  -- moment, definitions are only allowed at the top level,
+  -- so there can't be any inside the argument to build.
+  -- (Though perhaps there is an argument that this ought to be
+  -- relaxed specifically in the cases of 'Build' and 'Reprogram'.)
+  -- See #349
+  checkRequiredDevices ::
+    (Has (State GameState) sig m, Has (State Robot) sig m, Has (Error Exn) sig m) =>
+    Inventory ->
+    Term ->
+    Text ->
+    m [S.Set Entity]
+  checkRequiredDevices inventory cmd subject = do
+    currentContext <- use $ robotContext . defCaps
+    em <- use entityMap
+    creative <- use creativeMode
+    let -- Find out what capabilities are required by the program that will
+        -- be run on the target robot, and what devices would provide those
+        -- capabilities.
+        (caps, _capCtx) = requiredCaps currentContext cmd
+
+        -- list of possible devices per capability
+        capDevices = map (`deviceForCap` em) . S.toList $ caps
+
+        -- device is ok if it is available in the inventory of parent
+        -- when building or installed in target robot when reprogramming
+        deviceOK d = inventory `E.contains` d
+        ignoreOK ([], miss) = ([], miss)
+        ignoreOK (ds, _miss) = (ds, [])
+
+        (deviceSets, missingDeviceSets) =
+          Lens.over both (nubOrd . map S.fromList) . unzip $
+            map (ignoreOK . L.partition deviceOK) capDevices
+
+        formatDevices = T.intercalate " or " . map (^. entityName) . S.toList
+
+    -- check if robot has all devices to execute new command
+    (creative || all null missingDeviceSets)
+      `holdsOrFail` ( singularSubjectVerb subject "do" :
+                      "not have required devices:" :
+                      (("\n  - " <>) . formatDevices <$> filter (not . null) missingDeviceSets)
+                    )
+    -- check that there are in fact devices to provide every required capability
+    (creative || not (any null deviceSets))
+      `holdsOrFail` [singularSubjectVerb subject "can", "not perform an impossible task"]
```

Comment:
To make sure I'm understanding correctly, if this "impossible task" error is triggered, it means that there do not exist any devices to provide the required capability, right?  As in, we have some capability but no actual devices yet in `entities.yaml` which provide it?

Reply:
Hmm, I actually got this error while playing in classic mode and it's very mysterious.  How hard would it be to output the name of the capability for which there is no device, with a message apologizing and a link to the issue tracker?

Reply:
I expected this to only come up for `CGod` tier tasks, like `create "life"`, where the message would make sense.

Maybe I could check if it is `CGod` and otherwise link to the issue. Something like:
```Haskell
[ singularSubjectVerb subject "do", "not have", indefiniteQ capName
, "capability, because no device provides it."
, "\nSee https://github.com/swarm-game/swarm/issues/26."
]
```

However, I feel like it should be `Incapable` exceptions job to figure this out. :smirk:

Reply:
Some of it was common, but building and reprogramming can give a more specialised error message, so I keep some duplicity code. Oh well, at least I get to close #342 :innocent:

Reply:
Also, just an update: the reason I got this error while playing in classic mode was actually due to #389 .  So that was making it more mysterious than necessary.